frontend/src/app/feature/room/room.component.ts

```diff
@@ -39,131 +48,126 @@ export class RoomComponent extends WithWebSocket implements OnInit, OnDestroy {
   }
 
   ngOnInit() {
-    this.route.paramMap.subscribe(params => {
-      const roomId = params.get('roomId');
-
-      if (this.dashboardService.roomIds.indexOf(roomId) === -1) {
-        this.leaveRoom();
-        return;
-      }
-
-      this.room = new Room(roomId);
-
-      const myPlayer = this.dashboardService.getPlayer(roomId);
-      if (myPlayer) {
-        this.init(myPlayer);
-      } else {
-        setTimeout(() => { // hack to avoid 'value has been changed before check'
-          this.dialog.open(RoomDialogComponent).afterClosed().subscribe(this.onDialogClose(roomId));
-        });
-      }
-    });
-  }
-
-  private init(player: Player) {
-    this.you = player;
-    this.room.addPlayer(this.you);
-    this.connect(`${WsConstants.roomTopic}${this.room.id}`);
-  }
-
-  private onDialogClose(roomId: string) {
-    return (name: string) => {
-      if (!name) {
-        this.leaveRoom();
-      } else {
-        this.init(new Player(this.userService.userId, name));
-        this.dashboardService.addPlayer(roomId, this.you);
-      }
-    };
+    this.route.paramMap.pipe(
+      map(params => params.get('roomId')),
+      flatMap(roomId => this.roomService.getRoom(roomId)),
+      tap(room => this.throwIfEmpty(room)),
+      tap(room => {
+        room.players.map(p => Player.addColorIfMissing(p));
+        this.systemPlayer = Player.systemPlayer(room.id);
+        this.room = room;
+      }),
+      flatMap(room => this.playerService.findOrCreatePlayer(new Player(null, this.userService.id, room.id, this.userService.name))),
+      tap(player => {
+        this.currentPlayer = Player.newPlayer(player);
+        this.addPlayerTab(this.currentPlayer);
+      })
+    ).subscribe(
+      () => {
+        this.connect(`${ApiConst.WS_ROOM}/${this.room.id}`);
+      },
+      () => this.leaveRoom()
+    );
   }
 
   ngOnDestroy() {
-    if (this.wsConnection) {
-      this.disconnect();
-    }
+    this.leaveRoom();
   }
 
   onSendMessage() {
     if (this.message) {
-      this.sendMessage({ type: WsRoomMessageType.CHAT_MESSAGE, data: this.message });
+      this.notifyAll({ type: WsRoomMessageType.CHAT_MESSAGE, data: this.message });
       this.message = '';
     }
   }
 
-  getPlayers(): Array<Player> {
-    return this.room && Array.from<Player>(this.room.players.values())
-      .sort((a, b) => a.id === this.you.id ? -1 : 0);
-  }
-
-  onInventorySave = (inventory: Inventory) => this.sendMessage({ type: WsRoomMessageType.INVENTORY, data: inventory });
+  onInventorySave = (inventory: Inventory) => this.notifyAll({ type: WsRoomMessageType.INVENTORY, data: inventory });
 
   leaveRoom() {
     if (this.wsConnection) {
-      this.onClose();
+      this.onWsClose(); // hmm, looks like a hack
+    }
+    if (this.room && this.currentPlayer) {
+      this.roomService.removePlayerFromRoom(this.room.id, this.currentPlayer.id).subscribe(); // not sure about this, may not work stable
```

Comment:
i think it makes sense to leave player tab, so others can see what he has or smth =)

Reply:
Will think about it later